### PR TITLE
[core] Allow uv cache at installation

### PR DESCRIPTION
**Pull request by dentiny**
See issue: https://github.com/ray-project/ray/issues/49131

This PR provides an option to enable cache for `uv`.

### CODE REVIEW

File: python/ray/_private/runtime_env/uv.py
Comment (jjyao) [inline python/ray/_private/runtime_env/uv.py:176]: Is there a valid use case where we don't want cache?

Comment (dentiny) [reply python/ray/_private/runtime_env/uv.py:176]: Yes, cache lies in shared filesystem, managed by pip or uv.
The side-effect for cache is larger disk consumption.

Comment (sfriedowitz) [reply python/ray/_private/runtime_env/uv.py:176]: Another option that might be worth considering to add is the [link mode](https://docs.astral.sh/uv/reference/settings/#link-mode), which can have a pretty significant impact on cache installtion performance depending on filesystem characteristics. Although this can also be configured by an env_var, so I would understand not wanting to expand the number of arguments here.

Comment (jjyao) [reply python/ray/_private/runtime_env/uv.py]: `uv pip install --help` displays tons of options. Instead of explicitly add support for each of them, one way is that we can have a single `uv_pip_install_options` argument which is a list of strings (default is ["--no-cache"]), this way user has the most flexibility to specify whatever options they want. One downside is that user is more likely to shoot themselves in the foot but as long as we can surface good error messages it should be fine. WDYT @sfriedowitz @pcmoritz?

Comment (sfriedowitz) [reply python/ray/_private/runtime_env/uv.py]: That would work great for my use case!

Comment (dentiny) [reply python/ray/_private/runtime_env/uv.py]: Sounds good! I updated my implementation based on the suggestion.

File: doc/source/ray-core/handling-dependencies.rst
Comment (jjyao) [inline doc/source/ray-core/handling-dependencies.rst:412]: should be `List[str]`

Comment (dentiny) [reply doc/source/ray-core/handling-dependencies.rst:412]: There's no difference between these two types? We simply take whatever users provide.

Comment (dentiny) [reply doc/source/ray-core/handling-dependencies.rst:412]: A TODO on Hao: need to add a unit test and check how string works here.

Comment (dentiny) [reply doc/source/ray-core/handling-dependencies.rst:412]: @jjyao I think you're right, passing string values doesn't work. Updated to list[str].

File: python/ray/_private/runtime_env/uv.py
Comment (jjyao) [inline python/ray/_private/runtime_env/uv.py:169]:
```suggestion
        # 2. Allow user to specify their own options to install packages via `uv`.
```

File: python/ray/_private/runtime_env/validation.py
Comment (jjyao) [inline python/ray/_private/runtime_env/validation.py:182]:
```suggestion
                    "runtime_env['uv']['uv_pip_install_options'] must be of type List[str] "
                    ", "
```

File: python/ray/_private/runtime_env/validation.py
Comment (jjyao) [inline python/ray/_private/runtime_env/validation.py:191]:
```suggestion
                        "runtime_env['uv']['uv_pip_install_options'] must be of type "
                        "List[str], "
                        f"got {type(cur_opt)} for {i}th option"
```

**Review by jjyao — COMMENTED** [review]
LG

File: doc/source/ray-core/handling-dependencies.rst
Comment (dentiny) [reply doc/source/ray-core/handling-dependencies.rst:413]: Fixed.

File: python/ray/_private/runtime_env/validation.py
Comment (dentiny) [reply python/ray/_private/runtime_env/validation.py:182]: Updated.

File: python/ray/_private/runtime_env/validation.py
Comment (dentiny) [reply python/ray/_private/runtime_env/validation.py:191]: Updated.

File: python/ray/_private/runtime_env/uv.py
Comment (dentiny) [reply python/ray/_private/runtime_env/uv.py]: Updated.

File: python/ray/_private/runtime_env/uv.py
Comment (jjyao) [inline python/ray/_private/runtime_env/uv.py:181]:
```suggestion
        if uv_opt_list:
```

**Review by jjyao — APPROVED** [review]

File: python/ray/_private/runtime_env/uv.py
Comment (dentiny) [reply python/ray/_private/runtime_env/uv.py:181]: This is intentional, I try to avoid any implicit conversion.

Just an example why I don't like implicit conversion
```
if (!a_pointer) {
  // do something
}
```
I find it hard to read whether the branch is used for nullptr or not-null, so I prefer explicit and clear stratement.

Comment (dentiny) [reply python/ray/_private/runtime_env/uv.py:181]: We don't need to worry about the performance here, len is stored as a data member.

Comment (jjyao) [reply python/ray/_private/runtime_env/uv.py:181]: Yea, I don't worry about perf, it's just more pythonic: https://peps.python.org/pep-0008/

Comment (dentiny) [reply python/ray/_private/runtime_env/uv.py]: Updated anyway, btw what's the detailed item for this? There're so many items in the doc you link..

Comment (jjyao) [reply python/ray/_private/runtime_env/uv.py]:
```
For sequences, (strings, lists, tuples), use the fact that empty sequences are false:
# Correct:
if not seq:
if seq:
# Wrong:
if len(seq):
if not len(seq):
```

Comment (dentiny) [reply python/ray/_private/runtime_env/uv.py]: Thank you!

**Comment by dentiny** [comment]
Hi @jjyao , could you please help me merge the PR? Thank you!

**Comment by jjyao** [comment]
Waiting for approval from doc team.

File: python/ray/tests/test_runtime_env_uv.py
Comment (pcmoritz) [inline python/ray/tests/test_runtime_env_uv.py:38]: Hmm, this is actually part of the test, I think `pip_install_test` is a test package and this tests if the import works. The whole test seems to be using `pip` instead of `uv`, do you want to try to update that? Or remove the test altogether for uv, I'm not sure if it makes sense for uv. But before you remove it it would be great to first try to understand it (I don't currently).

**Review by pcmoritz — APPROVED** [review]

File: python/ray/tests/test_runtime_env_uv.py
Comment (pcmoritz) [reply python/ray/tests/test_runtime_env_uv.py:38]: Feel free to do this work in a follow up PR but please do follow up on the `import pip_install_test` removal, because without that the test makes no sense.

Comment (dentiny) [reply python/ray/tests/test_runtime_env_uv.py:38]: Thanks for sharing! I revert the change back.

**Pull request merged**
